tensorflow_recommenders_addons/dynamic_embedding/python/ops/shadow_embedding_ops.py

```diff
@@ -0,0 +1,213 @@
+
+# lint-as: python3
+"""
+Dynamic Embedding is designed for Large-scale Sparse Weights Training.
+See [Sparse Domain Isolation](https://github.com/tensorflow/community/pull/237)
+
+`The file will be introduced as `shadow_ops` under `dynamic_embedding`.
+It is a submodule of `dynamic_embedding`.
+
+In TensorFlow 2.x, tf.function is introduced to speedup the computaion.
+And also modular programming based on [tf.Module](https://www.tensorflow.org/guide/intro_to_modules)
+are recommended because of the Pythonic style APIs. But APIs like
+`embedding_lookup`, `embedding_lookup_unique`, `embedding_lookup_sparse`, and
+`safe_embedding_lookup_sparse` in `dynamic_embedding`, are wrappers of
+`embedding_lookup`. And it will create a TrainableWrapper object inside
+the function, which doesn't meet the requirements of
+[tf.function](https://www.tensorflow.org/guide/function)
+
+The `shadow_ops` submodule is designed to support usage on `tf.function`
+and modular style development, like keras.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import functools
+
+from tensorflow_recommenders_addons import dynamic_embedding as de
+
+from tensorflow.python.eager import context
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import ops
+from tensorflow.python.framework import tensor_shape
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import resource_variable_ops
+from tensorflow.python.training.tracking import base as trackable
+
+
+class ShadowVariable(de.TrainableWrapper):
+  """
+  ShadowVariable maps the sparse domain, which may reside cross multiple
+  devices, as a projection on current device. Its value represents the activated
+  part of the sparse domain. When lookup on sparse domain, it will fetch the
+  lookup result to local, and could be regarded as trainable object to
+  optimizers, like an ordinary variable. It supports the modular programming
+  and [tf.function](https://www.tensorflow.org/guide/function).
+  """
+
+  def __init__(self,
+               params,
+               name='ShadowVariable',
+               max_norm=None,
+               trainable=True,
+               **kwargs):
+    """
+    Create a ShadowVariable object.
+
+    Args:
+      params: A dynamic_embedding.Variable object represents the sparse domain.
+      ids: If set, it needs to be a ResourceVariable, to keep the
+        ids for backward computations. Otherwise the ShadowVariable will
+        create ids variable buffer itself.
+      name: Name of the ShadowVariable.
+      max_norm: If not `None`, each values is clipped if its l2-norm is larger
+        than this value.
+      trainable: Bool. If true, the variable will be treated as trainable.
+        Default is true.
+
+      **kwargs:
+        model_mode: ModelMode of the option. Default is ModelMode.CURRENT_SETTING.
```

Comment:
Why not make these definite inputs explicit？

Reply:
model_mode: I think it is set as a global setting. So the option rarely had chance to be set manually.
create_as_slot: Only for internal usage. TF does not allow multiple jumps for trackable slots currently. To avoid hack to internal codes to saveable mechanism, I use this flag to discriminate whether if the ShadowVariable is created as a slots.
ids: Same as above.
exists: Same as above.
distribute_strategy: Maybe it could be explicit.